python/src/scipp/compat/mantid.py

```diff
@@ -1,91 +1,108 @@
-    convert_instrument(d, ws)
-    initPosSpectrumNo(d, nHist, ws)
+    comp_info = convert_instrument(ws)
+    pos, num = initPosSpectrumNo(nHist, ws)
 
     # TODO More cases?
-    tag = sp.Coord.DeltaE if ws.getAxis(
-        0).getUnit().unitID() == 'DeltaE' else sp.Coord.Tof
-    dim = sp.Dim.DeltaE if ws.getAxis(
-        0).getUnit().unitID() == 'DeltaE' else sp.Dim.Tof
+    # tag = sc.Dim.EnergyTransfer if ws.getAxis(
+    #     0).getUnit().unitID() == 'DeltaE' else sc.Dim.Tof
+    dim = sc.Dim.EnergyTransfer if ws.getAxis(
+        0).getUnit().unitID() == 'DeltaE' else sc.Dim.Tof
     if cb:
-        d[tag] = ([dim], ws.readX(0))
+        coords = sc.Variable([dim], values=ws.readX(0))
     else:
-        d[tag] = ([sp.Dim.Position, dim],
-                  (ws.getNumberHistograms(), len(ws.readX(0))))
+        coords = sc.Variable([sc.Dim.Position, dim],
+                             shape=(ws.getNumberHistograms(),
+                                    len(ws.readX(0))))
         for i in range(ws.getNumberHistograms()):
-            d[tag][sp.Dim.Position, i] = ws.readX(i)
-
-    d[sp.Data.Value, name] = ([sp.Dim.Position, dim],
-                              (ws.getNumberHistograms(), len(ws.readY(0))))
-    d[sp.Data.Variance, name] = (
-        [sp.Dim.Position, dim], (ws.getNumberHistograms(), len(ws.readE(0))))
+            coords[sc.Dim.Position, i].values = ws.readX(i)
 
     # TODO Use unit information in workspace, if available.
-    d[sp.Data.Value, name].unit = sp.units.counts
-    d[sp.Data.Variance, name].unit = sp.units.counts * sp.units.counts
+    var = sc.Variable([sc.Dim.Position, dim],
+                      shape=(ws.getNumberHistograms(), len(ws.readY(0))),
+                      unit=sc.units.counts)
 
     for i in range(ws.getNumberHistograms()):
-        d[sp.Data.Value, name][sp.Dim.Position, i] = ws.readY(i)
-        d[sp.Data.Variance, name][sp.Dim.Position,
-                                  i] = ws.readE(i) * ws.readE(i)
+        var[sc.Dim.Position, i].values = ws.readY(i)
+        var[sc.Dim.Position, i].variances = ws.readE(i) * ws.readE(i)
 
-    return d
+    return sc.DataArray(data=var, coords={dim: coords, sc.Dim.Position: pos},
+                        labels={"spectrum_number": num,
+                                "component_info": comp_info})
 
 
-def ConvertEventWorkspaceToDataset(ws, name, drop_pulse_times):
-    d = sp.Dataset()
+def ConvertEventWorkspaceToDataset(ws, drop_pulse_times):
     nHist = ws.getNumberHistograms()
-    convert_instrument(d, ws)
-    initPosSpectrumNo(d, nHist, ws)
-
-    d[sp.Data.Events, name] = ([sp.Dim.Position], (nHist,))
-    for i, e in enumerate(d[sp.Data.Events, name].data):
-        e[sp.Data.Tof] = ([sp.Dim.Event], ws.getSpectrum(i).getTofs())
+    comp_info = convert_instrument(ws)
+    pos, num = initPosSpectrumNo(nHist, ws)
+
+    coords = sc.Variable([sc.Dim.Position, sc.Dim.Tof],
+                         shape=[nHist, sc.Dimensions.Sparse])
```

Comment:
Maybe we should throw if the X unit if not Tof? Some for Workspace2D above (throw if case not handled, i.e., do not default to Tof if it is not energy-transfer)?